### ISSUE TITLE
Formulation : Test 11.1.3 : Etiquette invisible

### ISSUE THREAD

**Issue by innoticFR** [DISIC/RGAA#55]
Le test 11.1.3 traite du cas particulier d'un champ de formulaire ayant une étiquette non visible.

Le WCAG [Understanding Success Criterion 3.3.2: Labels or Instructions](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html) oblige pour les champs de formulaire à la présence d'un libellé visible.

> NOTE: A label is presented to all users whereas the name may be hidden and only exposed by assistive technology. In many (but not all) cases the name and the label are the same.

Le paragraphe "Intent" évoque la problématique non couverte par le Test 11.3
>  It is possible for controls and inputs to have an appropriate accessible name or description (e.g. using `aria-label="..."`) and therefore pass Success Criterion 4.1.2, but to still fail this Success Criterion

voir [technique H44](https://www.w3.org/TR/WCAG20-TECHS/H44.html) également
> However, for Success Criterion 3.3.2, the label element must be visible since it provides assistance to all users who need help understanding the purpose of the field.

Seul le "nom accessible" (retransmis aux lecteurs d'écran, par exemple) peut être masqué, mais pas l'étiquette.

Les techniques proposées dans le RGAA pour rendre visible le libellé nécessitent la prise de focus (clavier ou souris), ce qui n'est pas l'intention du WCAG (navigation à la parole, par exemple).

Ce test reste toutefois très intéressant dans le cas de champs dont le libellé visible, mais pas à proximité, est référencé par `aria-labelledby`, et que l'utilisateur ne possède pas de lecteur d'écran (ex: loupe d'écran).

**Comment by juliemoynat-tanaguru** [comment]
Je pense qu'il s'agit du test 11.1.3 et non 11.3 ?

Je suis également gênée par ce test qui permet l'utilisation d'un attribut `title` si le libellé du champ est masqué alors qu'il ne s'affiche qu'à l'usage de la souris. En revanche, il ne permet pas que le libellé soit repris dans l'attribut `placeholder` du champ. D'un point de vue expérience utilisateur, c'est certes discutable mais je ne vois pas comment on pourrait permettre l'attribut `title` et refuser le `placeholder` dans le cas présent.

**Comment by innoticFR** [comment]
Merci pour la correction! C'était bien cela.

Au sujet du `placeholder` [le WAI indique](https://www.w3.org/WAI/tutorials/forms/instructions/#placeholder-text)
> While placeholder text provides valuable guidance for many users, placeholder text is not a replacement for labels. Assistive technologies, such as screen readers, do not treat placeholder text as labels. 

Il viendra donc en compléments d'autres techniques (par exemple un bouton "OK" dans le cadre d'un champ de recherche à champ unique ou un bouton en forme de loupe) et à condition d'avoir un nom accessible.

Le cas de mauvais usage typique est le cas où le `placeholder` est invisible parce que le champ est pré-rempli (ou auto-completion hasardeuse côté navigateur)

Dans un formulaire multi-champs, je ne vois pas d'exemple de cas où le placeholder pourrait être bénéficiable...

**Comment by juliemoynat-tanaguru** [comment]
Je ne dis pas que le `placeholder` viendrait en remplacement du libellé puisque dans le test 11.1.3, il est bien question du cas où le libellé (nom accessible) est caché (aria-label, `<label>` masqué via CSS, `aria-labelledby` non accolé). Donc le champ a bien un nom accessible. Cela ne vient donc pas à l'encontre de la citation de la WAI.

Cet attribut `placeholder` n'est pas plus mauvais que l'attribut `title` dans ce genre d'usage.

Bien sûr, dans un monde idéal, on met les libellés visibles partout. 

Mais sur le cas précis du test 11.1.3, l'attribut `title` ne sera pas vu en navigation tactile, en navigation clavier ou en navigation via pilotage à la voix. Le `placeholder`, oui. Donc, pourquoi ne pas l'autoriser ?

**Comment by innoticFR** [comment]
Oui dans tous les cas l'attribut `title` ne doit pas entrer en compte pour évaluer le critère 3.3.2 du WCAG. Je viens de rajouter dans mon ticket la note suivante du WCAG afin d'éviter la confusion entre **libellé** et **nom accessible**. 
> NOTE: A label is presented to all users whereas the name may be hidden and only exposed by assistive technology. In many (but not all) cases the name and the label are the same.

Le critère 3.3.2 parle du **libellé** visible, pas du **nom accessible**.

Il y a un bon article de Steve Faulkner au sujet du placeholder : [HTML5 Accessibility Chops: the placeholder attribute](https://developer.paciellogroup.com/blog/2011/02/html5-accessibility-chops-the-placeholder-attribute/) qui est effectivement une alternative, qui viendrait à mon sens en complément de l'attribut `title`. L'attribut `placeolder` entre bien dans le [calcul du **nom accessible**](https://www.w3.org/TR/html-aam-1.0/#input-type-text-input-type-password-input-type-search-input-type-tel-input-type-url-and-textarea-element) des balises `<input>`.

Le problème est de savoir si on peut le considérer comme un **libellé** alors qu'il est invisible quand le champ est rempli...

**Comment by alixraison** [comment]
Concernant le fait que l’étiquette d’un champ doive être forcément visible, le RGAA définit une étiquette comme un “texte à proximité du champ de formulaire permettant d'en connaître la nature, le type ou le format des informations attendues.” (https://www.numerique.gouv.fr/publications/rgaa-accessibilite/methode/glossaire/#etiquette-de-champ-de-formulaire). Cette définition ne se réduit pas au seul élément <label> comme c’est le cas de la technique H44. Et le SC 3.2.2 concerne globalement différentes manières de libeller un champ. La technique H44 reste une technique non normative.

Il reste tout à fait légitime d’envisager une étiquette non explicitement visible pourvu qu’une technique supplétive soit présente ; ces techniques sont certes discutables d’un point de vue utilisateur (attribut title, prise de focus) mais elles sont considérées comme valides du point de vue des WCAG. Ainsi, concernant l’usage de l’attribut title, il est justifié par la technique H65 qui fournit des exemples de cas d’utilisation (par exemple, un ensemble de champs composés pour la saisie d’une donnée).

Concernant l’attribut placeholder, il n’est pas considéré comme une technique possible pour jouer le rôle d’un libellé de substitution. On trouve une référence à l’attribut dans la technique F102 qui concerne le SC 1.4.10 (Reflow) pour illustrer un cas d’échec du critère qui ferait usage d’un tel attribut seul : “Labels above text inputs are hidden and replaced by placeholder text after reflow, without a technique showing dedicated labels when focusing the fields.” (https://www.w3.org/WAI/WCAG21/Techniques/failures/F102).

Aucune modification n’est envisagée.

**Comment by innoticFR** [comment]
Je ne suis pas d'accord. Une étiquette invisible n'est ni légitime ni valide dans le WCAG, et c'est bien là toute l’ambiguïté qu'il faut lever.

Il est à noter que la technique H65 n'apparait plus dans les techniques possibles pour le 3.3.2 : Label or Instructions
https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html 
voir  https://github.com/w3c/wcag/issues/101

Elle apparait toutefois au 4.1.2 : Name, Role, Value
https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html

**Comment by stevenMouret** [comment]
Le `title` n'est visible qu'au survol à la souris, en mobile par exemple l'attribut `title` n'est pas visible alors que l'attribut `placeholder` l'est.
Le placeholder a beaucoup de défaut, il est caché à la prise de focus, il ne peut pas être sur plusieurs lignes et doit être court. De plus la spec HTML 5.2 précise que c'est une aide et non un label. Sur un champ courriel par exemple c'est bien l'aide à la saisie qui doit apparaître (john.doe@mail.com) et non le type de champ attendu "Courriel".
Dans les deux cas, aucune solution n'est vraiment robuste cependant l'attribut `title` est conforme aux WCAG.

Remettre en cause l'attribut `title` sur ce critère c'est aussi le remettre en question sur les liens, les buttons, les images, les iframes... ce qui me semble très compliqué.

Par contre si on part du principe que l'attribut `title` n'est pas un élément visible (il ne l'est quand dans certaines conditions) alors le test 11.1.3 ne me semble pas correcte sur la condition du `title`.

Le RGAA se veut être conforme aux WCAG, supprimer le test du `title` n’empêcherait pas une conformité WCAG.

**Comment by innoticFR** [comment]
Encore une fois, il ne faut pas confondre le nom accessible et le libellé visible.

Il est tout à fait accepté qu'une image, un lien ou un bouton graphique ou une iframe n'ait pas un libellé visible. 

Le critère 3.3.2 traite exclusivement des contrôles qui nécessitent une saisie de l'utilisateur (*"when content requires user input."*) et traite explicitement du libellé visible dans ce cas précis. Donc exit définitif l'attribut `title` ou toute étiquette invisible. 

Il est nécessaire de lire attentivement le critère 3.3.2 pour savoir qu'évidemment un libellé ne peut pas être invisible:
> NOTE: **A label is presented to all users whereas the name may be hidden and only exposed by assistive technology**

Le but de ce libellé est de permettre notamment aux utilisateurs de technologies d'assistance d'accéder au champ. Un exemple typique est le cas d'un utilisateur navigant à la parole.

Concernant l'attribut `placeholder` (qui n'est pas le sujet de ce ticket), son utilisation seule [ne serait être suffisante ](https://github.com/w3c/wcag/issues/864) pour ce critère .

**Comment by stevenMouret** [comment]
> Le critère 3.3.2 traite exclusivement des contrôles qui nécessitent une saisie de l'utilisateur ("when content requires user input.") et traite explicitement du libellé visible dans ce cas précis. Donc exit définitif l'attribut title ou toute étiquette invisible.

Pourtant le dernier exemple de succès est basé sur l'attribut `title` sans libellés visibles.

> so invisible labels are provided with the "title" attribute to each of the three fields.

Donc selon les WCAG l'utilisation de l'attribut `title` seul est tout à fait conforme.

**Comment by innoticFR** [comment]
Non, il faut voir la phrase dans son ensemble : 
> **To address this, the three fields are grouped in a fieldset with the legend "Phone number"**. Visual labels for the fields (beyond the punctuation) cannot be provided in the design, so invisible labels are provided with the "title" attribute to each of the three fields.

Il y a déjà un libellé visible pour le groupe d'élément étant entendu que "*The term label is not limited to the label element in HTML.*" C'est juste que le libellé décrit l'ensemble des 3 champs. Donc le critère 3.3.2 est satisfait. Chaque élément a ensuite un nom accessible. 

On peut se référer à l'échec correspondant:
[Failure of Success Criterion 3.3.2 by visually formatting a set of phone number fields but not including a text label](https://www.w3.org/WAI/WCAG21/Techniques/failures/F82)

> However, even if all the fields have programmatically determined names, a text label must also identify the set of fields as a phone number.

La première condition porte explicitement:
> For each set of phone number fields in the web page that represents a single phone number, check that the set of fields are labeled with a **visible text label that is positioned near the set of phone number fields**.

Il y a un ticket ouvert sur le Github du WCAG qui soulève le problème de la mauvaise formulation et montre que l'écriture de cet exemple est contestée et pourrait générer quelques ambiguïtés si lue trop rapidement.
https://github.com/w3c/wcag/issues/755

**Comment by innoticFR** [comment]
Statut suite au passage au RGAA 4.1

Malgré le caractère majeur de cette problématique, aucune réponse n'a été apportée dans la dernière version du RGAA

**Comment by acaodinsic** [comment]
Le ticket concerne la version RGAA 4.0.

**design-gouv** y donne une réponse et n'envisage pas de modification.

Je ferme donc ce ticket qui concerne la version RGAA 4.0

Néanmoins, si le point donne encore lieu à des questions dans la version RGAA 4.1, il conviendra d'ouvrir un nouveau ticket pour la version 4.1 avec éventuellement une référence liée à ce ticket, pour qu'elles puissent être analysées lors d'une amélioration ultérieure. Le ticket pourrait, par exemple, proposer une formulation plus adaptée du test.
